Bug:  Gradle Plugin: compileTestJava fails when using overlay file for a class found in a bundle on the buildpath

## Summary

When moving to the bnd gradle plugin (biz.aQute.bnd:biz.aQute.bnd.gradle) 7.0.0 and higher, unit tests do not compile cleanly when the source file in a project overlays a file in a buildpath dependency.

For example, if the overlay file adds a new method that a unit test references, it fails to find it and gets a compile failure


---

## Environment

- **Bnd version:** 7.0.0+
- **Build tool:** Gradle
- **Java:** 17+

---

## Steps to Reproduce

1. Reference a group:artifact:version in your -buildpath
2. Overlay one of the classes from that dependency and add a new method
3. Include contents of the dependency in the bundle and add the overlay class on top of it using `-includeresource` or `Include-Resource`
4. Have a unit test that references the new method
5. Run `./gradlew build` or `./gradlew compileTestJava`
6. Observe java compile failures when building the unit test

---

## Root Cause

I don't know the cause for this particular problem.  It is as if the project's output directory or the built bundle for the project is not prepending to the build path.  Maybe it is appended or not added at all when doing unit test compilation.  Note that the Eclipse bndtools plugin does not have this problem.  It is appears to only be a problem with the gradle plugin.

---

### Workaround

To workaround the problem, I add the following to my gradle configuration.

```
compileTestJava {
  // With the 7.x versions of the bnd gradle plugin the test classpath for compiling
  // does not have the output directory of the source of the project so if there are
  // overlays that add new methods, the new methods are not found during test compile time.
  classpath = files(sourceSets.main.output.classesDirs) + classpath
}
```

Using similar set up to what I did in #7303, here are some steps to reproduce the problem:

### Directory structure
```
workspace/
├── cnf/
│   ├── bnd.bnd
│   ├── build.bnd
│   └── deps.maven
├── project/
│   ├── bnd.bnd
│   ├── build.gradle
│   ├── src/
│   │   └── org/objectweb/asm/
│   │       └── Attribute.java
│   └── test/
│       └── org/objectweb/asm/
│           └── AttributeTest.java
└── settings.gradle
```

### `settings.gradle`
```groovy
buildscript {
    repositories { maven { url = uri('https://repo1.maven.org/maven2') } }
    dependencies {
        classpath 'biz.aQute.bnd:biz.aQute.bnd.gradle:7.0.0'
    }
}

import aQute.bnd.build.Workspace
import aQute.bnd.osgi.Constants
import aQute.bnd.gradle.BndPlugin

Workspace.setDriver(Constants.BNDDRIVER_GRADLE)
Workspace.addGestalt(Constants.GESTALT_BATCH, null)
Workspace workspace = new Workspace(rootDir, 'cnf').setOffline(startParameter.offline)

workspace.getAllProjects().each { include it.name }

gradle.projectsLoaded { gradle ->
    gradle.rootProject.ext.bndPlugin = BndPlugin.class
    gradle.rootProject.ext.bndWorkspace = workspace
    gradle.rootProject.subprojects.each { subproject ->
        if (subproject.name != 'cnf') {
            subproject.apply plugin: BndPlugin
        }
    }
}
```

### `cnf/bnd.bnd`
```
-nobundles=true
```

### `cnf/build.bnd`
```
-plugin.1.Central: \
  aQute.bnd.repository.maven.provider.MavenBndRepository; \
    name = Central; \
    releaseUrl = https://repo.maven.apache.org/maven2/; \
    index = ${build}/deps.maven
```

### `cnf/deps.maven`
```
org.ow2.asm:asm:9.10
junit:junit:4.13.2
org.hamcrest:hamcrest-core:1.3
```

### `project/bnd.bnd`
```
Bundle-Version: 1.0.0

# Pull in everything from the upstream asm jar, then overlay our modified Attribute class on top
-includeresource: @${repo;org.ow2.asm:asm;9.10}!/*, \
                  org/objectweb/asm/Attribute.class=bin/org/objectweb/asm/Attribute.class

-buildpath: \
    org.ow2.asm:asm;version=9.10

-testpath: \
    junit:junit;version=4.13.2, \
    org.hamcrest:hamcrest-core;version=1.3
```

### `project/build.gradle`
```groovy
// Workaround for https://github.com/bndtools/bnd/issues/7306
// With the 7.x versions of the bnd gradle plugin the test classpath for compiling
// does not have the output directory of the source of the project so if there are
// overlays that add new methods, the new methods are not found during test compile time.
//
// Uncomment the block below to apply the workaround and see compileTestJava succeed.

/*compileTestJava {
    classpath = files(sourceSets.main.output.classesDirs) + classpath
}*/
```

### `project/src/org/objectweb/asm/Attribute.java`
```java
package org.objectweb.asm;

/**
 * Overlay of org.objectweb.asm.Attribute that adds a new helper method.
 * This class overlays the same class from the asm dependency on the -buildpath.
 * The new method getTypeName() does not exist in the upstream jar.
 */
public class Attribute extends Object {

    public final String type;

    protected Attribute(final String type) {
        this.type = type;
    }

    /**
     * New method added by overlay — does NOT exist in the upstream asm jar.
     * A unit test calling this method triggers the bug in the bnd Gradle plugin 7.0.0+:
     * compileTestJava cannot find this method because the project's own output directory
     * is not prepended to the test compile classpath.
     */
    public String getTypeName() {
        return type;
    }
}
```

### `project/test/org/objectweb/asm/AttributeTest.java`
```java
package org.objectweb.asm;

import org.junit.Test;
import static org.junit.Assert.assertEquals;

/**
 * Unit test that calls the new method added by the overlay class.
 * With the bnd Gradle plugin 7.0.0+, compileTestJava fails because
 * the project's own output directory is not on the test compile classpath,
 * so getTypeName() — which does not exist in the upstream asm jar — is not found.
 */
public class AttributeTest {

    @Test
    public void testGetTypeName() {
        Attribute attr = new Attribute("MyAttr") {};
        assertEquals("MyAttr", attr.getTypeName());
    }
}
```

### Run
```
./gradlew project:compileTestJava
```

## Expected

`compileTestJava` succeeds because `getTypeName()` is defined in the project's own compiled overlay class.

## Actual

```
> Task :project:compileJava
> Task :project:compileTestJava FAILED
  error: cannot find symbol
      assertEquals("MyAttr", attr.getTypeName());
                                 ^
    symbol:   method getTypeName()
    location: variable attr of type Attribute
```

`compileJava` succeeds (the overlay compiles fine), but `compileTestJava` fails because the project's `bin/` output directory is not prepended to the test compile classpath. The upstream `asm` jar's `Attribute` class — which has no `getTypeName()` method — is resolved instead of the overlay.

## Workaround

Uncomment the block in `project/build.gradle`:

```groovy
compileTestJava {
    classpath = files(sourceSets.main.output.classesDirs) + classpath
}
```

--- a/gradle-plugins/biz.aQute.bnd.gradle/src/main/java/aQute/bnd/gradle/BndPlugin.java
+++ b/gradle-plugins/biz.aQute.bnd.gradle/src/main/java/aQute/bnd/gradle/BndPlugin.java
@@ -238,7 +238,7 @@ public class BndPlugin implements Plugin<Project> {
 			/* Set up source sets */
 			SourceSetContainer sourceSets = sourceSets(project);
 			/* bnd uses the same directory for java and resources. */
-			sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME, sourceSet -> {
+			SourceSet mainSourceSet = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME, sourceSet -> {
 				ConfigurableFileCollection srcDirs;
 				try {
 					srcDirs = objects.fileCollection()
@@ -249,25 +249,23 @@ public class BndPlugin implements Plugin<Project> {
 				File destinationDir = bndProject.getSrcOutput();
 				sourceSet.getJava()
 					.setSrcDirs(srcDirs);
+				sourceSet.getJava()
+					.getDestinationDirectory()
+					.fileValue(destinationDir);
 				sourceSet.getResources()
 					.setSrcDirs(srcDirs);
-				sourceSet.getJava()
+				sourceSet.getResources()
 					.getDestinationDirectory()
 					.fileValue(destinationDir);
 				sourceSet.getOutput()
 					.setResourcesDir(destinationDir);
-				TaskProvider<AbstractCompile> compileTask = tasks.named(sourceSet.getCompileJavaTaskName(),
-					AbstractCompile.class, t -> {
-						t.getDestinationDirectory()
-							.fileValue(destinationDir);
-						FileCollection jarLibraryElements = jarLibraryElements(t, sourceSet.getCompileClasspathConfigurationName());
-						t.setClasspath(jarLibraryElements.plus(t.getClasspath()));
-					});
-				generateInputAction.ifPresent(compileTask::configure);
-				sourceSet.getOutput()
-					.dir(Maps.of("builtBy", compileTask.getName()), destinationDir);
-				TaskProvider<Task> processResourcesTask = tasks.named(sourceSet.getProcessResourcesTaskName());
-				generateInputAction.ifPresent(processResourcesTask::configure);
+				sourceSet.setCompileClasspath(jarLibraryElements(project, sourceSet.getCompileClasspathConfigurationName()));
+				sourceSet.setRuntimeClasspath(objects.fileCollection()
+					.from(sourceSet.getOutput(), jarLibraryElements(project, sourceSet.getRuntimeClasspathConfigurationName())));
+				generateInputAction.ifPresent(action -> {
+					tasks.named(sourceSet.getCompileJavaTaskName()).configure(action);
+					tasks.named(sourceSet.getProcessResourcesTaskName()).configure(action);
+				});
 			});
 			sourceSets.getByName(SourceSet.TEST_SOURCE_SET_NAME, sourceSet -> {
 				ConfigurableFileCollection srcDirs = objects.fileCollection()
@@ -275,22 +273,20 @@ public class BndPlugin implements Plugin<Project> {
 				File destinationDir = bndProject.getTestOutput();
 				sourceSet.getJava()
 					.setSrcDirs(srcDirs);
+				sourceSet.getJava()
+					.getDestinationDirectory()
+					.fileValue(destinationDir);
 				sourceSet.getResources()
 					.setSrcDirs(srcDirs);
-				sourceSet.getJava()
+				sourceSet.getResources()
 					.getDestinationDirectory()
 					.fileValue(destinationDir);
 				sourceSet.getOutput()
 					.setResourcesDir(destinationDir);
-				TaskProvider<AbstractCompile> compileTask = tasks.named(sourceSet.getCompileJavaTaskName(),
-					AbstractCompile.class, t -> {
-						t.getDestinationDirectory()
-							.fileValue(destinationDir);
-						FileCollection jarLibraryElements = jarLibraryElements(t, sourceSet.getCompileClasspathConfigurationName());
-						t.setClasspath(jarLibraryElements.plus(t.getClasspath()));
-					});
-				sourceSet.getOutput()
-					.dir(Maps.of("builtBy", compileTask.getName()), destinationDir);
+				sourceSet.setCompileClasspath(objects.fileCollection()
+					.from(mainSourceSet.getOutput(), jarLibraryElements(project, sourceSet.getCompileClasspathConfigurationName())));
+				sourceSet.setRuntimeClasspath(objects.fileCollection()
+					.from(sourceSet.getOutput(), mainSourceSet.getOutput(), jarLibraryElements(project, sourceSet.getRuntimeClasspathConfigurationName())));
 			});
 			/* Configure srcDirs for any additional languages */
 			project.afterEvaluate(p -> {
@@ -305,26 +301,20 @@ public class BndPlugin implements Plugin<Project> {
 								sourceDirectorySets.put(name, sourceDirectorySet);
 							}
 						});
-					Provider<Directory> destinationDir = sourceSet.getJava()
-						.getClassesDirectory();
 					TaskProvider<Task> processResourcesTask = tasks.named(sourceSet.getProcessResourcesTaskName());
 					sourceDirectorySets.forEach((name, sourceDirectorySet) -> {
 						try {
 							TaskProvider<AbstractCompile> compileTask = tasks.named(sourceSet.getCompileTaskName(name),
-								AbstractCompile.class, t -> {
-									t.getDestinationDirectory()
-										.value(destinationDir);
-									t.getInputs()
-										.files(processResourcesTask)
-										.withPropertyName(processResourcesTask.getName());
-								});
-							generateInputAction.ifPresent(compileTask::configure);
+								AbstractCompile.class, t -> t.getInputs()
+									.files(processResourcesTask)
+									.withPropertyName(processResourcesTask.getName())
+							);
 							sourceDirectorySet.setSrcDirs(sourceSet.getJava()
 								.getSrcDirs());
 							sourceDirectorySet.getDestinationDirectory()
-								.value(destinationDir);
-							sourceSet.getOutput()
-								.dir(Maps.of("builtBy", compileTask.getName()), destinationDir);
+								.value(sourceSet.getJava()
+									.getDestinationDirectory());
+							generateInputAction.ifPresent(compileTask::configure);
 						} catch (UnknownDomainObjectException e) {
 							// no such task
 						}
@@ -341,25 +331,19 @@ public class BndPlugin implements Plugin<Project> {
 								sourceDirectorySets.put(name, sourceDirectorySet);
 							}
 						});
-					Provider<Directory> destinationDir = sourceSet.getJava()
-						.getClassesDirectory();
 					TaskProvider<Task> processResourcesTask = tasks.named(sourceSet.getProcessResourcesTaskName());
 					sourceDirectorySets.forEach((name, sourceDirectorySet) -> {
 						try {
 							TaskProvider<AbstractCompile> compileTask = tasks.named(sourceSet.getCompileTaskName(name),
-								AbstractCompile.class, t -> {
-									t.getDestinationDirectory()
-										.value(destinationDir);
-									t.getInputs()
-										.files(processResourcesTask)
-										.withPropertyName(processResourcesTask.getName());
-								});
+								AbstractCompile.class, t -> t.getInputs()
+									.files(processResourcesTask)
+									.withPropertyName(processResourcesTask.getName())
+							);
 							sourceDirectorySet.setSrcDirs(sourceSet.getJava()
 								.getSrcDirs());
 							sourceDirectorySet.getDestinationDirectory()
-								.value(destinationDir);
-							sourceSet.getOutput()
-								.dir(Maps.of("builtBy", compileTask.getName()), destinationDir);
+								.value(sourceSet.getJava()
+									.getDestinationDirectory());
 						} catch (UnknownDomainObjectException e) {
 							// no such task
 						}
@@ -376,11 +360,7 @@ public class BndPlugin implements Plugin<Project> {
 			/* Set up dependencies */
 			DependencyHandler dependencies = project.getDependencies();
 			dependencies.add(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME, pathFiles(bndProject.getBuildpath()));
-			dependencies.add(JavaPlugin.RUNTIME_ONLY_CONFIGURATION_NAME, objects.fileCollection()
-				.from(bndProject.getSrcOutput()));
 			dependencies.add(JavaPlugin.TEST_IMPLEMENTATION_CONFIGURATION_NAME, pathFiles(bndProject.getTestpath()));
-			dependencies.add(JavaPlugin.TEST_RUNTIME_ONLY_CONFIGURATION_NAME, objects.fileCollection()
-				.from(bndProject.getTestOutput()));
 			/* Set up compile tasks */
 			ConfigurableFileCollection javacBootclasspath = objects.fileCollection()
 				.from(decontainer(bndProject.getBootclasspath()));

--- a/gradle-plugins/biz.aQute.bnd.gradle/src/main/java/aQute/bnd/gradle/BndUtils.java
+++ b/gradle-plugins/biz.aQute.bnd.gradle/src/main/java/aQute/bnd/gradle/BndUtils.java
@@ -158,12 +158,11 @@ public class BndUtils {
 	 * attribute, the {@code jar} variant will also be available, making any resources packaged in the jar
 	 * visible.  The provided configuration itself will <strong>NOT</strong> be modified.
 	 *
-	 * @param task The Task.
+	 * @param project The Project.
 	 * @param configurationName The configuration name.
 	 * @return A {@link FileCollection} of the {@code jar} variants of the specified configuration.
 	 */
-	public static FileCollection jarLibraryElements(Task task, String configurationName) {
-		Project project = task.getProject();
+	public static FileCollection jarLibraryElements(Project project, String configurationName) {
 		Configuration configuration = project.getConfigurations()
 			.getByName(configurationName);
 		ArtifactView artifactView = configuration.getIncoming()
@@ -172,10 +171,10 @@ public class BndUtils {
 				try {
 					attributes.attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, project.getObjects()
 						.named(LibraryElements.class, LibraryElements.JAR));
-					task.getLogger()
+					project.getLogger()
 						.info("Using jars for configuration {}:{}", project.getPath(), configurationName);
 				} catch (IllegalArgumentException e) {
-					task.getLogger()
+					project.getLogger()
 						.info("Unable to use jars for configuration {}:{}", project.getPath(), configurationName, e);
 				}
 			});

--- a/gradle-plugins/biz.aQute.bnd.gradle/src/main/java/aQute/bnd/gradle/BundleTaskExtension.java
+++ b/gradle-plugins/biz.aQute.bnd.gradle/src/main/java/aQute/bnd/gradle/BundleTaskExtension.java
@@ -209,7 +209,7 @@ public class BundleTaskExtension {
 		outputDirectory = objects.directoryProperty();
 		SourceSet mainSourceSet = sourceSets(project).getByName(SourceSet.MAIN_SOURCE_SET_NAME);
 		setSourceSet(mainSourceSet);
-		classpath(jarLibraryElements(task, mainSourceSet.getCompileClasspathConfigurationName()));
+		classpath(jarLibraryElements(project, mainSourceSet.getCompileClasspathConfigurationName()));
 		properties = objects.mapProperty(String.class, Object.class)
 			.convention(project.provider(this::getGradleProjectProperties));
 		defaultBundleSymbolicName = task.getArchiveBaseName()

